### PR TITLE
Update examples and fixed pruning error

### PR DESCRIPTION
**Pull request by PenghuiCheng**
# What does this PR do?
This PR will adapt to new API of neural_compressor, and Fixed pruning sparsity issue.

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you make sure to update the documentation with your changes?
- [ ] Did you write any new necessary tests?

### CODE REVIEW

File: examples/neural_compressor/language-modeling/run_clm.py [hunk 1]
```diff
@@ -591,16 +593,22 @@ def compute_metrics(eval_preds):
 
     if optim_args.apply_pruning:
 
-        if optim_args.end_epoch is None:
-            end_epoch = training_args.num_train_epochs
+        if optim_args.end_step is None:
+            end_step = training_args.num_train_epochs * (
```
Comment (echarlaix) [inline examples/neural_compressor/language-modeling/run_clm.py:597]: Shouldn't we include the number of `gradient_accumulation_steps` ? Also I think the number of epoch might be easier for users to set, what do you think ?

File: tests/neural_compressor/test_optimization.py [hunk 2]
```diff
@@ -229,7 +237,7 @@ def compute_metrics(p: EvalPrediction):
                 transformers_outputs = transformers_model(**tokens)
             self.assertTrue(torch.allclose(onnx_outputs.logits, transformers_outputs.logits, atol=1e-4))
             sparsity = trainer.get_model_sparsity()
-            self.assertGreaterEqual(sparsity, target_sparsity * 100)
+            self.assertGreaterEqual(sparsity, target_sparsity * 100 / 2)
```
Comment (echarlaix) [inline tests/neural_compressor/test_optimization.py:240]: Where does the  `/ 2` comes from ?

**Review by echarlaix — COMMENTED** [review]
Thanks for adding this @PenghuiCheng !

File: examples/neural_compressor/language-modeling/run_clm.py [hunk 3]
```diff
@@ -157,17 +157,19 @@ class OptimizationArguments:
...
         default=None,
-        metadata={"help": "Epoch for which the pruning process will end."},
+        metadata={"help": "step for which the pruning process will end."},
     )
     pruning_approach: str = field(
-        default="basic_magnitude",
-        metadata={"help": "Pruning approach. Supported approach is basic_magnitude."},
+        default="snip_momentum",
+        metadata={
+            "help": "Pruning approach. Supported approach is 'magnitude'," "'gradient', 'snip_momentum', and 'snip'."
```
[the remark this answers is not in the document]
Comment (PenghuiCheng) [reply examples/neural_compressor/language-modeling/run_clm.py]: Added the pruning method introduction in README and optimization_inc.mdx

File: examples/neural_compressor/language-modeling/run_clm.py [hunk 4]
```diff
@@ -591,16 +593,22 @@ def compute_metrics(eval_preds):
...
+                optim_args.end_step,
+                training_args.num_train_epochs * (len(train_dataset) // training_args.per_device_train_batch_size),
+            )
 
         pruning_config = WeightPruningConfig(
-            start_epoch=optim_args.start_epoch,
-            end_epoch=end_epoch,
+            start_step=optim_args.start_step,
+            end_step=end_step,
+            excluded_op_names=["classifier"],
```
[the remark this answers is not in the document]
Comment (PenghuiCheng) [reply examples/neural_compressor/language-modeling/run_clm.py]: Yes, excluded_op_names is for improving accuracy in some cases. I think we can remove it in examples. Users can ensure which op to be excluded.

File: examples/neural_compressor/language-modeling/run_clm.py (hunk 1 above)
Comment (PenghuiCheng) [reply examples/neural_compressor/language-modeling/run_clm.py:597]: Pruning occurs after the every optimizer.step() is called. So I think needn't include gradient_accumulation_steps. Our think the step can slightly adjust sparsity.

File: tests/neural_compressor/test_optimization.py (hunk 2 above)
Comment (PenghuiCheng) [reply tests/neural_compressor/test_optimization.py:240]: "get_model_sparsity()" return the sparsity of the whole model, in the model, no all ops will be pruned, there are many op unpruned, and target_sparsity is only indicates that the sparsity for the op that can do prune. In this case, if we set the target sparsity to 0.9,  the model sparsity is about 0.56, So I divide the result by 2.

File: examples/neural_compressor/language-modeling/run_clm.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (echarlaix) [reply examples/neural_compressor/language-modeling/run_clm.py]: Yes sure we can remove it!

File: tests/neural_compressor/test_optimization.py (hunk 2 above)
Comment (PenghuiCheng) [reply tests/neural_compressor/test_optimization.py:239]: @echarlaix, I add the explanation of pruning in the README and docs/source/optimization_inc.mdx. 
Thanks!

File: examples/neural_compressor/language-modeling/run_clm.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (echarlaix) [reply examples/neural_compressor/language-modeling/run_clm.py]: Thank you !

File: tests/neural_compressor/test_optimization.py (hunk 2 above)
Comment (echarlaix) [reply tests/neural_compressor/test_optimization.py:239]: Thank you @PenghuiCheng

**Review by echarlaix — APPROVED** [review]
